### ISSUE TITLE
Replace keyframes concept with better waypoint controls

### ISSUE THREAD

**Issue by mray** [synfig/synfig#265]
I see fundamental flaws in the keyframes concept as Synfig uses them now:

1. keyframes make waypoints always subject to move, placing a waypoint is never a guarantee that it stays there 
2. keyframes limit making use of different waypoint types in one frame
3. keyframes can create unnecessary/unwanted extra waypoints
4. they are a complex concept with lots of potential of misunderstanding and misuse

I admit that I don't have experience, and maybe I'm just missing some information. Most benefits of keyframes could be achieved with better waypoint controls in my eyes. Right now I don't want to suggest a particular approach, but hope to get feedback on this issue to finally have create a proposal.

I'm particularly interested in how people currently make use of keyframes, what the specific reason/benefit is of using them.

**Comment by caryoscelus** [comment]
1. that sounds more like a feature
2. ??
3. true, but if you don't want those waypoints, you should perhaps not create kf there

The greatest thing about keyframes is that they provide some functionality, but are entirely optional. It would be great to have better alternative for some of kf usecases, though it might not necessarily replace them, but complement instead.

**Comment by mray** [comment]
> The greatest thing about keyframes is that they provide some functionality, but are entirely optional.

What is the functionality?
If the functionality is essential, how can it be optional?

**Comment by morevnaproject** [comment]
Hello @mray ,

> I don't have experience, and maybe I'm just missing some information

I am sorry to say that, but I  think this is exactly the case.

> I'm particularly interested in how people currently make use of keyframes, what the specific reason/benefit is of using them. 

The best explanation with usage examples could be found in the official Training Course, lesson 5 - http://www.synfig.org/cms/en/news/new-version-synfig-training-course/

You might also want to look at https://www.youtube.com/watch?v=05grnEPMKlg and http://wiki.synfig.org/Keyframe

In short:

- Waypoint remembers value for particular parameter of particular layer at given time.
- Keyframe is used to "remember" (fixate) ALL parameters for ALL layers in compositions.

Also, 

- Moving Keyframe results in moving all waypoints, which belong to it. That's a feature. It also moves in-between waypoints. That's a feature. That way you can easily adjust timing of your scene without dealing with each individual waypoins.

So, keyframe is high-level tool to control low-level objects (waypoints and parameters).

Keyframes are intended for 2 main purposes:

- With keyframe  you can be confident that your composition at certain moment of time remains as it is. This is important and makes Synfig unique. For example, in Moho (Anime Studio) there is no such guarantee, and you can end up tweaking your animation here and there.
- With keyframe it is easy to tune timing of your animation.

> If the functionality is essential, how can it be optional?

Keyframes ARE optional. It's a powerful feature, but you are free NOT to use it. Just use waypoints if you comfortable with this. To be honest, when I have started to learn Synfig in 2008, I was also avoided keyframes, until I figured out their advantages.

Again, my advice to  watch videos how people working with Synfig. You can find some recorded streams on our channel - https://www.youtube.com/c/MorevnaprojectOrg

**Comment by mray** [comment]
By now I have a pretty good concept of keyframes in Synfig I think.
To me it looks like keyframes *do* solve many problems with waypoints, but they also introduce new ones.

So, unfortunately I am *not* free to *"just NOT use them"*: 
– **without** keyframes it is hard to have decent control of waypoints
– **with** keyframes you lose proper control of waypoints

Maybe this is not true for everybody, but for my workflow (motion graphics) Synfigs combination of waypoints/keyframes is not really feasible. Synfig is the most capable 2D motion graphics tool in free software, I think. So there is much to be lost if it unusable as such.

**Comment by caryoscelus** [comment]
I think discussion would be more constructive if you'd make a list of specific things you want regarding waypoint controlling. Then either someone will point out how to do that via existing functionality or we'll have a specific feature request.

**Comment by mray** [comment]
AfterEffects has a [good approach](https://youtu.be/pTGViL5v8cA?t=257). 

Specifically:
* [Keyframe handle controls](https://youtu.be/pTGViL5v8cA?t=556)
* [Rubber band selection](https://youtu.be/pTGViL5v8cA?t=610)
* [Copy & Paste](https://youtu.be/pTGViL5v8cA?t=613)
* [Roving Keyframes](https://youtu.be/sIT7MDz0iAE?t=34)
* [Parameter Filter](https://youtu.be/pTGViL5v8cA?t=491) with shortcuts
: S=scale P=position...

**Comment by mray** [comment]
Accidental early post. but anyway, I know it is kind of lame to point at Adobe at all the things they have to offer, but this way of dealing with keyframes is the direction i'm talking about.

**Comment by caryoscelus** [comment]
Well, these are unrelated to keyframes. And most of them were kinda mentioned in previous issue.

Anyway, i think it makes sense to create individual issues for specific features/feature groups.

Some notes:

**Roving keyframes**
While there's no special gui for speed and such, you can get similar effect using "free time" layer (or time dilation parameter in group). Or perhaps that would be overkill, because when you want to follow a path, you'll have a parameter responsible for movement progress - you can animate it however you like.

**Copy/paste**
while you can't really copy&paste, you can duplicate waypoints in two ways:

- duplicate individual wp by rmb->duplicate when time cursor is at another frame
- duplicate selected wps by shift-dragging them

**Comment by caryoscelus** [comment]
#266 #267 #268

**Comment by mray** [comment]
I think this need clarification. This seems to be the main benefit of keyframes:

   > *  With keyframe you can be confident that your composition at certain moment of time remains as it is. This is important and makes Synfig unique. For example, in Moho (Anime Studio) there is no such guarantee, and you can end up tweaking your animation here and there.
   > *  With keyframe it is easy to tune timing of your animation.

* Having confidence in a composition remaining fixed is one of the reasons I'm proposing this; keyframes mess with waypoints by moving them without consent. If it wasn't for keyframes I don't see how compositions can get messed. If something need to be fixed: add waypoints.

* Having Rubberband selection (+ scaling of waypoint groups), and handle controls covers "easy time tuning" even better than keyframes can. Those two basic concepts alone are invaluable to fine tune timing, and I can't imagine how keyframes compare to that.

I'm happy to get educated about how keyframes are a good general approach, but currently I can only see its benefit due to primitive waypoint handling in Synfig.

@morevnaproject Lesson 5 seems to point to a paywall so I didn't get a look at that.

**Comment by caryoscelus** [comment]
How to mess up animation in three easy steps:

1. create an object and tweek its parameters to match first frame
2. switch to animate mode and set parameters on another frame
3. return to first frame and notice parameters the same as your second frame

This is the simplest way; in a more complex situation you might e.g. want to animate parameter you previously left static and not notice losing your static setup until it's too late.

Also i'm not sure this is the right place to ask for help - perhaps forum is a better choice.

**Comment by mray** [comment]
Oh, you mean if you start by removing the initially created keyframe this happens!
(which is odd to start with imho)
Enable animate mode *before* animating! Afaik this is how the buttons is supposed to work.

otherwise you could have an easy "insert keyframe" button for properties, like the stopwatch icon in AE or pressing "I" in blender. So you can pin down the state of everything at any time by only creating waypoints.

Keeping the discussion here keeps the context right, so I'm ok with that.

**Comment by morevnaproject** [comment]
@mray 

> Having confidence in a composition remaining fixed is one of the reasons I'm proposing this; keyframes mess with waypoints by moving them without consent.

Keyframes don't move waypoints without consent. If you move keyframe, then:
- All waypoints that belongs to this keyframe are moved with this keyframe
- All waypoints that belong to other keyframes remain in place.
- All waypoints that are **between** of moved and neighbor (static) keyframes are stretched according to keyframe movement.
- All other waypoints remain in place.

> Having Rubberband selection (+ scaling of waypoint groups), and handle controls covers "easy time tuning"

I agree that the waypoints manipulation system is not perfect. To be more clear - it is not finished. The named features are must have for sure. You might be interested to look here - http://www.synfig.org/forums/viewtopic.php?f=14&t=5013

> If it wasn't for keyframes I don't see how compositions can get messed up.

The detailed example is given in Training Course lesson 5, as I mentioned.

> @morevnaproject Lesson 5 seems to point to a paywall so I didn't get a look at that.

It's OK. Just consider, that this "paywall" helped to develop many features of Synfig during last 3 years. ^__^
If this is not an option for you, then you can read a transcription of lesson 5, you can get it here -     http://sync.morevnaproject.org/seafile/d/add8a5941f/

**Comment by mray** [comment]
> Keyframes don't move waypoints without consent. If you move keyframe, then:
> * All waypoints that belongs to this keyframe are moved with this keyframe
> * All waypoints that belong to other keyframes remain in place.
> * All waypoints that are between of moved and neighbor (static) keyframes are stretched according to keyframe movement.
> * All other waypoints remain in place.

Yes. This is how keyframes work! They influence all surrounding waypoints without you having a say in it. You cant decide on a waypoint level if it should be influenced or not. It is like forcing every keyframe to be part of a group, but the group also influences groups around it. 
So yes, you can move every waypoint to any location, but not without affecting its surrounding. But that is exactly the control I need and expect from Synfig.

**Comment by morevnaproject** [comment]
@mray If you don't want waypoint to be affected by movement of neighbor keyframes, then place a keyframe to it.

**Comment by mray** [comment]
@morevnaproject my problem with keyframes having too broad influence on the animation is not solved by adding more keyframes.
